src/hotspot/share/classfile/modules.cpp

```diff
@@ -605,6 +607,100 @@ void Modules::serialize(SerializeClosure* soc) {
+  if (soc->reading()) {
+    bool disable = false;
+    if (_archived_addmods_names[0] != '\0') {
+      if (Arguments::addmods_count() == 0) {
+        log_info(cds)("--add-modules module name(s) found in archive but not specified during runtime: %s",
+            _archived_addmods_names);
+        disable = true;
+      } else {
+        const char* addmods_names = get_addmods_names_as_sorted_string();
+        if (strcmp((const char*)_archived_addmods_names, addmods_names) != 0) {
+          log_info(cds)("Mismatched --add-modules module name(s).");
+          log_info(cds)("  dump time: %s runtime: %s", _archived_addmods_names, addmods_names);
+          disable = true;
+        }
+      }
+    } else {
+      if (Arguments::addmods_count() > 0) {
+        log_info(cds)("--add-modules module name(s) specified during runtime but not found in archive: %s",
+                      get_addmods_names_as_sorted_string());
+        disable = true;
+      }
+    }
+    if (disable) {
+      log_info(cds)("Disabling optimized module handling");
+      CDSConfig::stop_using_optimized_module_handling();
+    }
+    log_info(cds)("optimized module handling: %s", CDSConfig::is_using_optimized_module_handling() ? "enabled" : "disabled");
+    log_info(cds)("full module graph: %s", CDSConfig::is_using_full_module_graph() ? "enabled" : "disabled");
+  }
+}
+
+const char* Modules::get_addmods_names_as_sorted_string() {
+  ResourceMark rm;
+  const int max_digits = 3;
+  const int extra_symbols_count = 2; // includes '.', '\0'
+  size_t prop_len = strlen("jdk.module.addmods") + max_digits + extra_symbols_count;
+  char* prop_name = resource_allocate_bytes(prop_len);
+  GrowableArray<const char*> list;
+  for (unsigned int i = 0; i < Arguments::addmods_count(); i++) {
+    jio_snprintf(prop_name, prop_len, "jdk.module.addmods.%d", i);
+    const char* prop_value = Arguments::get_property(prop_name);
+    char* p = resource_allocate_bytes(strlen(prop_value) + 1);
+    strcpy(p, prop_value);
+    while (*p == ',') p++; // skip leading commas
+    while (*p) {
+      char* next = strchr(p, ',');
+      if (next == nullptr) {
+        // no more commas, p is the last element
+        list.append(p);
+        break;
+      } else {
+        *next = 0;
+        list.append(p);
+        p = next + 1;
+      }
+    }
+  }
+
+  // Example:
+  // --add-modules=java.compiler --add-modules=java.base,java.base,,
+  //
+  // list[0] = "java.compiler"
+  // list[1] = "java.base"
+  // list[2] = "java.base"
+  // list[3] = ""
+  // list[4] = ""
+  list.sort(ClassLoaderExt::compare_module_names);
+
+  const char* prefix = "";
+  stringStream st;
+  const char* last_string = ""; // This also filters out all empty strings
+  for (int i = 0; i < list.length(); i++) {
+    const char* m = list.at(i);
+    if (strcmp(m, last_string) != 0) { // filter out duplicates
+      st.print("%s%s", prefix, m);
+      last_string = m;
+      prefix = ",";
```

Comment:
You use comma as the separator.  Folks will have various opinions, but I think you should use a character which is positively forbidden by the JVMS 4.2.3.  That says, basically, that any character is fine (in the classfile format) EXCEPT control characters.  The most visible control character is probably `'\n'` so I suggest using that instead of comma.  Your key-strings will look like groups of sorted lines, one module name per line.  That's plenty readable.

If you keep comma, nothing will break, unless some smart guy figures out how to cook up a classfile that has a module name in it with a comma inside, that looks like two other legitimate modules names separated by that comma.  Something odd could happen next.  Better to avoid it from the start by using a truly illegal module name character.

(If these were class names I'd be suggesting `';'` or `'.'` not `'\n'` because the rules for class names are different, in the JVMS.  Again, none of these issues come up with inputs which are compiled from normal Java sources, but they can appear with craftily crafted classfiles.)

https://docs.oracle.com/javase/specs/jvms/se23/html/jvms-4.html#jvms-4.2.3

Reply:
Thanks for taking a look, John.
I've made the change based on your suggestion.